dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/security/config/OAuth2ConfirmAccessController.java

```diff
@@ -0,0 +1,138 @@
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.webapi.security.config;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.hisp.dhis.system.velocity.VelocityManager;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.security.oauth2.common.exceptions.OAuth2Exception;
+import org.springframework.security.oauth2.provider.AuthorizationRequest;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.SessionAttributes;
+import org.springframework.web.servlet.ModelAndView;
+import org.springframework.web.servlet.View;
+import org.springframework.web.util.HtmlUtils;
+
+@Controller
+@SessionAttributes( "authorizationRequest" )
+public class OAuth2ConfirmAccessController
+{
+    @Autowired
+    @Qualifier( "org.hisp.dhis.system.velocity.VelocityManager" )
+    VelocityManager velocityManager;
+
+    @RequestMapping( "/oauth/confirm_access" )
+    public ModelAndView getAccessConfirmationB( Map<String, Object> model, HttpServletRequest request )
+        throws Exception
+    {
+        Map<String, Object> vars = new HashMap<>();
+
+        AuthorizationRequest authorizationRequest = (AuthorizationRequest) model.get( "authorizationRequest" );
+        if ( authorizationRequest != null )
+        {
+            String clientId = authorizationRequest.getClientId();
+            vars.put( "client_id", clientId );
+        }
+
+        String approvalContent = velocityManager.render( vars, "confirm_access" );
+
+        if ( request.getAttribute( "_csrf" ) != null )
+        {
+            model.put( "_csrf", request.getAttribute( "_csrf" ) );
+        }
+
+        View approvalView = new View()
+        {
+            @Override
+            public String getContentType()
+            {
+                return "text/html";
+            }
+
+            @Override
+            public void render( Map<String, ?> model, HttpServletRequest request, HttpServletResponse response )
+                throws Exception
+            {
+                response.setContentType( getContentType() );
+                response.getWriter().append( approvalContent );
+            }
+        };
```

Comment:
This is used two times, could be extracted. Maybe there even already exists a view subclass allowing this?